src/qwircl/core.cljs

```diff
@@ -78,14 +83,78 @@
 
 (defn empty-location? [[x y] grid]
   (let [location (get-in grid [x y])]
-    (and 
+    (or
+     true
      (nil? (:color location))
      (nil? (:shape location)))))
 
-(defn valid-play? [state clicked]
-  (and 
-   (empty-location? clicked (:grid state))
-   true))
+(defn same-line? [positions [x y]]
+  (or (empty? positions)
+      (every? #(= (first (:coordinates %)) x)
+              positions)
+      (every? #(= (second (:coordinates %)) y) 
+              positions)))
+
+(def transformations {:horizontal {:x [dec inc] :y [identity identity]}
+                      :vertical {:x [identity identity] :y [dec inc]}})
+(defn reduce-neighbors 
+  ([grid [x y] f g h initial]
+   (loop [i (f x)
+          j (g y)
+          acc initial]
+     (if-let [cell (get-in grid [i j])]
+       (recur (f i) (g j) (h acc cell))
+       acc)))
+  ([grid clicked h initial direction]
+   (let [f1 (get-in transformations [direction :x 0])
+         f2 (get-in transformations [direction :x 1])
+         g1 (get-in transformations [direction :y 0])
+         g2 (get-in transformations [direction :y 1])]
+     (reduce-neighbors grid clicked f1 g1 h 
+                       (reduce-neighbors grid clicked f2 g2 h initial)))))
+
+(defn same-shape? [grid clicked tile direction]
+  (reduce-neighbors grid clicked #(and %1 (= (:shape tile) (:shape %2))) true direction))
+
+(defn same-color? [grid clicked tile direction]
+  (reduce-neighbors grid clicked #(and %1 (= (:color tile) (:color %2))) true direction))
+
+(defn touches-some-tile? [[x y] grid]
+  (let [uc (get-in grid [x (dec y)])
+        dc (get-in grid [x (inc y)])
+        lc (get-in grid [(dec x) y])
+        rc (get-in grid [(inc x) y])]
+    (or uc dc lc rc)))
+
+(defn all-unique? [grid clicked tile direction]
+  (let [ts (reduce-neighbors grid clicked conj [tile] direction)]
+    (= (count ts)
+       (count (set ts)))))
+
+(defn grid-with-positions [grid positions hand]
+  (reduce #(assoc-in %1 (:coordinates %2) (hand (first (:hand %2)))) grid positions))
+
+(defn valid-play? 
+  [{:keys [grid] {:keys [positions hand]} :pda :as state} clicked]
+  (let [previous-grid (grid-with-positions grid positions (get-in state [:my :hand]))
+        new-grid (grid-with-positions grid 
+                                      (conj positions {:coordinates clicked
+                                                       :hand (peek hand)})
+                                      (get-in state [:my :hand]))
+        tile (get-in state [:my :hand (first (peek hand))])]
```

Comment:
`(get-in state [my hand])` might be better destructured like `{{my-hand :hand} :my :as state}` or as its own let binding.